src/radiation_system.hpp

```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 				fourPiB = chat * a_rad * std::pow(T_gas, 4);
+				auto B = fourPiB / (4.0 * pi);
+				kappaPVec = ComputePlanckOpacity(rho, T_gas);
+				kappaFVec = ComputeFluxMeanOpacity(rho, T_gas);
+				kappaEVec = kappaFVec; // TODO(CCH): later, define ComputeEnergyMeanOpacity
 
 				// compute derivatives w/r/t T_gas
-				const double dB_dTgas = (4.0 * fourPiB) / T_gas;
-				const double dkappa_dTgas = RadSystem<problem_t>::ComputePlanckOpacityTempDerivative(rho, T_gas);
+				const double dB_dTgas = (4.0 * B) / T_gas;
 
 				// compute residuals
-				rhs = dt * (rho * kappa) * (fourPiB - chat * Erad_guess);
-				F_G = (Egas_guess - Egas0) + ((c / chat) * rhs);
-				F_R = (Erad_guess - Erad0) - (rhs + Src);
+				quokka::valarray<amrex::Real, nGroups_> radEnergyFractions{};
+				// TODO(CCH): fix this: if constexpr (nGroups_ > 1)
+				// if constexpr (nGroups_ > 1) {
+				// ../src/radiation_system.hpp(1089): error: An extended __device__ lambda cannot first-capture variable in constexpr-if context
+				// detected during: ...
+				if (nGroups_ > 1) {
+					radEnergyFractions = ComputePlanckEnergyFractions(radBoundaries_g, T_gas);
+					AMREX_ASSERT(min(radEnergyFractions) > 0.);
+				} else {
+					radEnergyFractions[0] = 1.0;
+				}
+
+				Rvec = dt * rho * (kappaPVec * fourPiB * radEnergyFractions - chat * kappaEVec * EradVec_guess);
+				Rtot = sum(Rvec);
+				F_G = Egas_guess - Egas0 + (c / chat) * Rtot;
+				F_R = EradVec_guess - Erad0Vec - (Rvec + Src);
 
 				// check if converged
-				if ((std::abs(F_G / Etot0) < resid_tol) && (std::abs(((c / chat) * F_R) / Etot0) < resid_tol)) {
+				if ((std::abs(F_G / Etot0) < resid_tol) && ((c / chat) * max(abs(F_R)) / Etot0 < resid_tol)) {
 					break;
 				}
 
-				// compute Jacobian elements
-				const double c_v = quokka::EOS<problem_t>::ComputeEintTempDerivative(rho, T_gas, massScalars);
-
-				drhs_dEgas = (rho * dt / c_v) * (kappa * dB_dTgas + dkappa_dTgas * (fourPiB - chat * Erad_guess));
+				quokka::valarray<double, nGroups_> dkappaP_dTgas = ComputePlanckOpacityTempDerivative(rho, T_gas);
+				quokka::valarray<double, nGroups_> dkappaE_dTgas = dkappaP_dTgas;
+				// TODO(CCH): later, define ComputeEnergyMeanOpacityTempDerivative
+				AMREX_ASSERT(!std::isnan(sum(dkappaP_dTgas)));
 
-				// Update variables
-				dFG_dEgas = 1.0 + (c / chat) * drhs_dEgas;
-				dFG_dErad = dt * (-(rho * kappa) * c);
-				dFR_dEgas = -drhs_dEgas;
-				dFR_dErad = 1.0 + dt * ((rho * kappa) * chat);
-				eta = -dFR_dEgas / dFG_dEgas;
-				eta = (eta > 0.0) ? eta : 0.0;
-
-				deltaErad = -(F_R + eta * F_G) / (dFR_dErad + eta * dFG_dErad);
-				deltaEgas = -(F_G + dFG_dErad * deltaErad) / dFG_dEgas;
+				// prepare to compute Jacobian elements
+				const double c_v = quokka::EOS<problem_t>::ComputeEintTempDerivative(rho, T_gas, massScalars);
+				dRtot_dErad = -dt * rho * kappaEVec * chat;
+				// TODO(CCH): fix this: dB_dTgas * radEnergyFractions is inaccurate. Define ComputePlanckEnergyFractionsTempDerivative
```

Comment:
Can you explain this?

Reply:
To calculate d B_i / d Tgas accurately, I need to consider that both B_i and radFractions_i is a function of Tgas. Currently I am using an approximation that d B_i / d Tgas = radFractions_i * d B / d Tgas, which is exact in 1-group case and mostly accurate in multigropu case. The error caused by this is below 0.2% in the RadTube test. I'll do it rigorously in the next PR.